relayer/src/chain/cosmos.rs

```diff
@@ -1259,109 +1287,70 @@ fn packet_from_tx_search_response(
     chain_id: &ChainId,
     request: &QueryPacketEventDataRequest,
     seq: Sequence,
-    mut response: tendermint_rpc::endpoint::tx_search::Response,
+    response: ResultTx,
 ) -> Option<IbcEvent> {
-    assert!(
-        response.txs.len() <= 1,
-        "packet_from_tx_search_response: unexpected number of txs"
-    );
-    if let Some(r) = response.txs.pop() {
-        let height = ICSHeight::new(chain_id.version(), u64::from(r.height));
-        if height > request.height {
-            return None;
-        }
-
-        let mut matching = Vec::new();
-        for e in r.tx_result.events {
-            if e.type_str != request.event_id.as_str() {
-                continue;
-            }
+    let height = ICSHeight::new(chain_id.version(), u64::from(response.height));
+    if request.height != ICSHeight::zero() && height > request.height {
+        return None;
+    }
 
-            let res = ChannelEvents::try_from_tx(&e);
-            if res.is_none() {
-                continue;
-            }
-            let event = res.unwrap();
-            let packet = match &event {
+    response
+        .tx_result
+        .events
+        .into_iter()
+        .filter(|abci_event| abci_event.type_str == request.event_id.as_str())
+        .flat_map(|abci_event| ChannelEvents::try_from_tx(&abci_event))
+        .find(|event| {
```

Comment:
Nitpick: we could split the call to `find` into a `flat_map` and a `find` (or `filter`+`last`, see my other comment) to make smaller logic units.